### ISSUE TITLE
Rstudio Viewer window not closed on shiny::stopApp()

### ISSUE THREAD

**Issue by vnijs** [rstudio/shiny#2033]
@kevinushey A while back I posted an [issue/question on Rstudio Community](https://community.rstudio.com/t/rstudio-viewer-window-not-closed-on-shiny-stopapp/4158) about Rstudio Viewer (and Rstudio Window) not (always) closing when closing a shiny app. I now have a minimal reproducible example (see code below). If you start the app and click the `Stop` button the app stops and the window closes. However, if you click the `Refresh` button **first** and then click on the `Stop` button the window does not close, even though the shiny app does stop. The error that pops-up from `shinyapp.js` is shown in the screen shot below. 

It seems that the Rstudio browser tries to reset the port when the `Refresh` button is pressed and the code in `shinyapp.js` is not able to capture that.

I think this might be an issue in Rstudio because I don't see it in Chrome or other browsers. I'm posting here but cc-ing @kevinushey (and @jcheng5 and @bborgesr)

This is an important issue in all my apps (e.g., [radiant.data](https://github.com/radiant-rstats/radiant.data/tree/MRB)). Being able to refresh the (Rstudio) browser from code is key to (fully) (re)setting the state of an app (i.e., data, inputs, etc.) based on a previous state loaded by the user.

`[Error] Unable to post message to http://127.0.0.1:4535. Recipient has origin http://127.0.0.1:25108.`

![image](https://user-images.githubusercontent.com/2998956/39086940-c4348c12-454d-11e8-9213-c5ac76c14c68.png)

```
ui <- fluidPage(
  actionButton("refresh", "Refresh", icon = icon("refresh")),
  uiOutput("refresh_window"),
  actionButton("stop", "Stop", class = "btn-danger", onclick = "setTimeout(function(){window.close();}, 100);")
)

server <- function(input, output) {
  observeEvent(input$stop, {
    stopApp(message("App stopped"))
  })

  output$refresh_window <- renderUI({
    req(input$refresh)
    tags$script("window.location.reload();")
  })
}

shinyApp(ui = ui, server = server)
```

![image](https://user-images.githubusercontent.com/2998956/39086866-961622c8-454d-11e8-8469-f34fa48d01bc.png)

**Comment by jcheng5** [comment]
The problem appears to be this:
https://github.com/rstudio/shiny/blob/9c3a0c86ca8f74b5a8ff562584240fde31d2e651/srcjs/shinyapp.js#L154

`document.referrer` starts out as the URL for RStudio itself, but after reload, it becomes the URL for the Shiny app. So in the latter case we call `postMessage` with the wrong origin. I guess we need to rely on something other than `document.referrer`.

**Comment by jcheng5** [comment]
Alternatively, we could decide that the fact that the app is disconnected is not privileged information and `postMessage('disconnected', '*')`.

**Comment by wch** [comment]
@jcheng5 I think using `'*'` is OK here.

**Comment by jcheng5** [comment]
@vnijs, my PR fixes it, but it's too late for Shiny v1.1--we'll merge it as soon as the master branch opens back up for Shiny v1.2. In the meantime, you can work around this issue by including the following JS snippet:

```R
tags$script("$(document).on('shiny:disconnected', function() {window.parent.postMessage('disconnected', '*');});")
```

Thanks for another great bug report!

**Comment by vnijs** [comment]
Wonderful! The work around works great! Thanks so much @jcheng5
